### PR TITLE
Port DeeplabV3Plus guide to Keras 3

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
This PR ports KerasCV's DeeplabV3Plus guide to Keras 3. Note that `MeanIoU` has been removed from the list of metrics when training the model since it errors out for the JAX backend due to dynamic shape error in Keras 3.

<details>
<summary>The error traceback</summary>

```
/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/core.py:64: UserWarning: Explicitly requested dtype int64 requested in array is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/google/jax#current-gotchas for more.
  return jnp.array(x, dtype=dtype)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-13-e5cb65984c58>](https://localhost:8080/#) in <cell line: 2>()
      1 keras.config.disable_traceback_filtering()
----> 2 model.fit(train_ds, validation_data=eval_ds, epochs=EPOCHS)

14 frames
[/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py](https://localhost:8080/#) in error_handler(*args, **kwargs)
    112     def error_handler(*args, **kwargs):
    113         if not is_traceback_filtering_enabled():
--> 114             return fn(*args, **kwargs)
    115 
    116         filtered_tb = None

[/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py](https://localhost:8080/#) in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq)
    440                 )
    441                 data = self._distribute_data(data)
--> 442                 logs, state = self.train_function(state, data)
    443                 (
    444                     trainable_variables,

    [... skipping hidden 12 frame]

[/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py](https://localhost:8080/#) in compiled_train_step(state, data)
    253             @partial(jax.jit, donate_argnames="state")
    254             def compiled_train_step(state, data):
--> 255                 return train_step(state, data)
    256 
    257             self.train_function = compiled_train_step

[/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py](https://localhost:8080/#) in one_train_step(state, data)
    234         def one_train_step(state, data):
    235             data = data[0]
--> 236             return self.train_step(state, data)
    237 
    238         def multi_train_steps(state, data):

[/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py](https://localhost:8080/#) in train_step(self, state, data)
    135         ) as scope:
    136             self._loss_tracker.update_state(unscaled_loss)
--> 137             logs = self.compute_metrics(x, y, y_pred, sample_weight)
    138 
    139         new_metrics_variables = []

[/usr/local/lib/python3.10/dist-packages/keras/src/trainers/trainer.py](https://localhost:8080/#) in compute_metrics(***failed resolving arguments***)
    366         del x  # The default implementation does not use `x`.
    367         if self._compile_metrics is not None:
--> 368             self._compile_metrics.update_state(y, y_pred, sample_weight)
    369         return self.get_metrics_result()
    370 

[/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py](https://localhost:8080/#) in update_state(self, y_true, y_pred, sample_weight)
    329         for m, y_t, y_p in zip(self._flat_metrics, y_true, y_pred):
    330             if m:
--> 331                 m.update_state(y_t, y_p)
    332         if sample_weight is not None:
    333             sample_weight = self._flatten_y(sample_weight)

[/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py](https://localhost:8080/#) in update_state(self, y_true, y_pred, sample_weight)
     16     def update_state(self, y_true, y_pred, sample_weight=None):
     17         for m in self.metrics:
---> 18             m.update_state(y_true, y_pred, sample_weight=sample_weight)
     19 
     20     def reset_state(self):

[/usr/local/lib/python3.10/dist-packages/keras/src/metrics/iou_metrics.py](https://localhost:8080/#) in update_state(self, y_true, y_pred, sample_weight)
    125         sample_weight = ops.cast(sample_weight, dtype=self.dtype)
    126 
--> 127         current_cm = confusion_matrix(
    128             y_true,
    129             y_pred,

[/usr/local/lib/python3.10/dist-packages/keras/src/metrics/metrics_utils.py](https://localhost:8080/#) in confusion_matrix(labels, predictions, num_classes, weights, dtype)
    684     values = ops.cast(values, dtype=dtype)
    685     num_classes = ops.cast(num_classes, "int64")
--> 686     confusion_matrix = ops.scatter(indices, values, (num_classes, num_classes))
    687     return confusion_matrix
    688 

[/usr/local/lib/python3.10/dist-packages/keras/src/ops/core.py](https://localhost:8080/#) in scatter(indices, values, shape)
     60     if any_symbolic_tensors((indices, values, shape)):
     61         return Scatter().symbolic_call(indices, values, shape)
---> 62     return backend.core.scatter(indices, values, shape)
     63 
     64 

[/usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/core.py](https://localhost:8080/#) in scatter(indices, values, shape)
    250 
    251 def scatter(indices, values, shape):
--> 252     zeros = jnp.zeros(shape, values.dtype)
    253     key = tuple(jnp.moveaxis(indices, -1, 0))
    254     return zeros.at[key].add(values)

[/usr/local/lib/python3.10/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in zeros(shape, dtype)
   2189     raise TypeError("expected sequence object with len >= 0 or a single integer")
   2190   dtypes.check_user_dtype_supported(dtype, "zeros")
-> 2191   shape = canonicalize_shape(shape)
   2192   return lax.full(shape, 0, _jnp_dtype(dtype))
   2193 

[/usr/local/lib/python3.10/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in canonicalize_shape(shape, context)
     79     return core.canonicalize_shape((shape,), context)  # type: ignore
     80   else:
---> 81     return core.canonicalize_shape(shape, context)  # type: ignore
     82 
     83 # Common docstring additions:

[/usr/local/lib/python3.10/dist-packages/jax/_src/core.py](https://localhost:8080/#) in canonicalize_shape(shape, context)
   2049   except TypeError:
   2050     pass
-> 2051   raise _invalid_shape_error(shape, context)
   2052 
   2053 def canonicalize_dim(d: DimSize, context: str="") -> DimSize:

TypeError: Shapes must be 1D sequences of concrete values of integer type, got (Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=1/0)>, Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=1/0)>).
If using `jit`, try using `static_argnums` or applying `jit` to smaller subfunctions.
The error occurred while tracing the function compiled_train_step at /usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py:253 for jit. This value became a tracer due to JAX operations on these lines:

  operation a:f32[] = convert_element_type[new_dtype=float32 weak_type=False] b
    from line /usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/core.py:64 (convert_to_tensor)
The error occurred while tracing the function compiled_train_step at /usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/trainer.py:253 for jit. This value became a tracer due to JAX operations on these lines:

  operation a:f32[] = convert_element_type[new_dtype=float32 weak_type=False] b
    from line /usr/local/lib/python3.10/dist-packages/keras/src/backend/jax/core.py:64 (convert_to_tensor)
```

</details>

@fchollet I think the above error is because of this offending line in Keras 3 that convert the `num_classes` argument to a tensor: https://github.com/keras-team/keras/blob/9c675a9a45e5e8244163fea82efc6066722608a1/keras/metrics/metrics_utils.py#L685

@mattdangerw Can you please help render the ipynb and md files? Thanks!

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! I am fixing the IoU bug.

**Comment by fchollet** [comment]
I have fixed the issue at HEAD. Please check that it works. If it works, then we should keep IoU (the example will then only work with the next Keras release, which should be soon).

File: guides/keras_cv/semantic_segmentation_deeplab_v3_plus.py [hunk 1]
```diff
@@ -46,20 +46,25 @@ class label such as person, bike, or background to each individual pixel of an
...
 ```
-%env KERAS_BACKEND=tensorflow
+import os
+
+os.environ["KERAS_BACKEND"] = "jax"
 ```
 """
 
-%env KERAS_BACKEND=tensorflow
+import os
```
Comment (mattdangerw) [inline guides/keras_cv/semantic_segmentation_deeplab_v3_plus.py]: I'm removing this, we shouldn't need it if the guide runs on any backend.

File: guides/keras_cv/semantic_segmentation_deeplab_v3_plus.py [hunk 2]
```diff
@@ -46,20 +46,25 @@ class label such as person, bike, or background to each individual pixel of an
 """
 
 """shell
-!pip install -q keras-core
-!pip install -q git+https://github.com/keras-team/keras-cv.git
+pip install -Uq keras-cv
```
Comment (mattdangerw) [inline guides/keras_cv/semantic_segmentation_deeplab_v3_plus.py]: I've been using

pip install -q --upgrade keras-cv
pip install -q --upgrade keras  # Upgrade to Keras 3.

No strong preference, just keeping this for consistency.

Comment (mattdangerw) [reply guides/keras_cv/semantic_segmentation_deeplab_v3_plus.py]: I added this while rendering.

**Comment by mattdangerw** [comment]
Added rendered files! Fix seemed to work, at least hit no snags running. Thanks @tirthasheshpatel !

**Comment by tirthasheshpatel** [comment]
> I have fixed the issue at HEAD. Please check that it works. If it works, then we should keep IoU (the example will then only work with the next Keras release, which should be soon).

I tried to run the notebook with keras install from source and everything works perfectly now. Thanks for the fix!

If we add it right now, the notebook won't render with the pypi version of Keras. Until then, if you are okay with the source install, I will re-add the MeanIoU part. What do you think?

**Comment by tirthasheshpatel** [comment]
Ah, sounds good @fchollet! Will push the changes today so we can merge!

**Review by pcoet — APPROVED** [review]

**Comment by pcoet** [comment]
@tirthasheshpatel Were you able to add the note Francois requested: "note that it will only be compatible with `keras>=3.0.1`"? I believe after that's added (and the .ipynb and .md files regenerated), we can go ahead and merge. Thanks!

**Comment by tirthasheshpatel** [comment]
Keras 3.0.1 already branched so no need for the note!

**Review by fchollet — APPROVED** [review]
LGTM, thank

**Pull request merged**